### ISSUE TITLE
Binary operators don't clear state of other binary operators

### ISSUE THREAD

**Issue by garlic-hub** [Arsenic-ATG/Qt-calculator#18]
**Describe the bug**
If you misclick a binary operator and attempt to click a different one, it will create two active binary operators and the equals is likely to get confused

**To Reproduce**
1. Press 5
2. Press +
3. Press x
4. Press 5
5. Press =

**Expected behavior**
The answer should be 25, but it says 10

**Comment by Arsenic-ATG** [comment]
confirmed 

Thanks for filing this bug.

**Comment by garlic-hub** [comment]
I know quite a bit of C++ myself and was just kind of bored looking for things to do. Anything you want help with?

**Comment by Arsenic-ATG** [comment]
> I know quite a bit of C++ myself and was just kind of bored looking for things to do. Anything you want help with?

It's basically because of the fact that when a binary operator button is pressed, the program just switches it's state to checked ( you can see `MainWindow::binary_operation_pressed()` for details ). 

And when performing the actual operation itself ( on the press of `equal` button ), it is just looking for a "checked" button and perform the operation depending on the first one it finds. 

**TL;DR** the program makes a wrong assumption that only one binary_operation button would be active at any given point.

**Comment by Arsenic-ATG** [comment]
If you want to work on it then just let me know, I will assign you this issue :)

Currently the idea I have is to reset all the buttons to "unchecked" state as soon a binary operation button is pressed, and then let is carry on its own process as usual 

so the `MainWindow::binary_operation_pressed` should look something like this : 
```c++
void MainWindow::binary_operation_pressed()
{
    // Only one binary operation can be active at a time.
    ui->pushButton_add->setChecked(false);
    ui->pushButton_minus->setChecked(false);
    ui->pushButton_multiply->setChecked(false);
    ui->pushButton_divide->setChecked(false);

    QPushButton* button = (QPushButton *)sender();
    //double labelnumber;
    //QString input;

    firstNum = ui->label->text().toDouble();
    button->setChecked(true);
}
```

I am interesting in listening to your thoughts for the same.

**Comment by garlic-hub** [comment]
I see an open issue to make a scientific version of the calculator. I don't think this will scale very well because every new binary oepration will add another if statement in `on_pushButton_equals_released()` and `setChecked(false)` in multiple places. I think a different approach would be to store the most recent pressed binary operation as state within the MainWindow class. Likely as an enum. This can easily be cleared and you can use a switch statement on it.

**Comment by Arsenic-ATG** [comment]
> I see an open issue to make a scientific version of the calculator. I don't think this will scale very well because every new binary oepration will add another if statement in `on_pushButton_equals_released()` and `setChecked(false)` in multiple places. I think a different approach would be to store the most recent pressed binary operation as state within the MainWindow class. Likely as an enum. This can easily be cleared and you can use a switch statement on it.

Indeed, that is just a quick patch that I proposed. 
I actually started working on this project over a year ago, and just like the other projects I started then, this one also didn't involved any kind of planning before.
I am thinking to restructure the entire program properly and adding things like creating a proper class hierarchy for different operators ( currently they are not related to each other at all ), and ( like you said ) performing operations with the help of most recent operation button pressed by the user.

**Comment by Arsenic-ATG** [comment]
Another approach was to take all the input as a string and parse the string to evaluate the expression at once, like this program would also be capable of calculating longer expressions and possibly throw errors for syntactically wrong expressions.

**Comment by Arsenic-ATG** [comment]
Sorry for getting side tracked 😅

> I see an open issue to make a scientific version of the calculator. I don't think this will scale very well because every new binary oepration will add another if statement in `on_pushButton_equals_released()` and `setChecked(false)` in multiple places. I think a different approach would be to store the most recent pressed binary operation as state within the MainWindow class. Likely as an enum. This can easily be cleared and you can use a switch statement on it.

Yes, this is much better way to do the same.
I would start working on it as soon as I get a chance to do so. 
Alternatively, if you are planning to contribute to the project, then just feel free to open a PR with the changes you like.

Thanks 🙂

**Comment by Arsenic-ATG** [comment]
Should be fixed after merge of PR #19
